### PR TITLE
[REEF-496] Adding Org.Apache.REEF.Evaluator.Test

### PR DESCRIPTION
**Pull request by jwang98052**
This PR
*created new test project for Evaluator
*moves evaluator tests from REEF.Tests to the new project
*Updated and added tests

JIRA: REEF-496(https://issues.apache.org/jira/browse/REEF-496)

This closes #

Author: Julia Wang  Email: juliaw@apache.org

### CODE REVIEW

File: lang/cs/Org.Apache.REEF.Evaluator.Tests/Org.Apache.REEF.Evaluator.Tests.csproj [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    <RootNamespace>Org.Apache.REEF.Evaluator.Tests</RootNamespace>
+    <AssemblyName>Org.Apache.REEF.Evaluator.Tests</AssemblyName>
+    <TargetFrameworkVersion>v4.5</TargetFrameworkVersion>
+    <FileAlignment>512</FileAlignment>
+    <SolutionDir Condition="$(SolutionDir) == '' Or $(SolutionDir) == '*Undefined*'">..</SolutionDir>
+    <RestorePackages>true</RestorePackages>
+    <ProjectTypeGuids>{3AC096D0-A1C2-E12C-1390-A8335801FDAB};{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}</ProjectTypeGuids>
+    <VisualStudioVersion Condition="'$(VisualStudioVersion)' == ''">10.0</VisualStudioVersion>
+    <VSToolsPath Condition="'$(VSToolsPath)' == ''">$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)</VSToolsPath>
+    <ReferencePath>$(ProgramFiles)\Common Files\microsoft shared\VSTT\$(VisualStudioVersion)\UITestExtensionPackages</ReferencePath>
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Evaluator.Tests/Org.Apache.REEF.Evaluator.Tests.csproj]: Is this path safe to be shared between us?

File: lang/cs/Org.Apache.REEF.Evaluator.Tests/Org.Apache.REEF.Evaluator.Tests.csproj [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+under the License.
+-->
+<Project ToolsVersion="12.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <PropertyGroup>
+    <ProjectGuid>{0B596FAC-53B0-435A-ACE5-6725309E414A}</ProjectGuid>
+    <OutputType>Library</OutputType>
+    <AppDesignerFolder>Properties</AppDesignerFolder>
+    <RootNamespace>Org.Apache.REEF.Evaluator.Tests</RootNamespace>
+    <AssemblyName>Org.Apache.REEF.Evaluator.Tests</AssemblyName>
+    <TargetFrameworkVersion>v4.5</TargetFrameworkVersion>
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Evaluator.Tests/Org.Apache.REEF.Evaluator.Tests.csproj]: Unrelated to this PR, but it just occurred to me: Shouldn't this be defined in build.props?

File: lang/cs/Org.Apache.REEF.Evaluator.Tests/Org.Apache.REEF.Evaluator.Tests.csproj [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    <ProjectTypeGuids>{3AC096D0-A1C2-E12C-1390-A8335801FDAB};{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}</ProjectTypeGuids>
+    <VisualStudioVersion Condition="'$(VisualStudioVersion)' == ''">10.0</VisualStudioVersion>
+    <VSToolsPath Condition="'$(VSToolsPath)' == ''">$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)</VSToolsPath>
+    <ReferencePath>$(ProgramFiles)\Common Files\microsoft shared\VSTT\$(VisualStudioVersion)\UITestExtensionPackages</ReferencePath>
+    <IsCodedUITest>False</IsCodedUITest>
+    <TestProjectType>UnitTest</TestProjectType>
+  </PropertyGroup>
+  <Import Project="$(SolutionDir)\build.props" />
+  <PropertyGroup>
+    <BuildPackage>false</BuildPackage>
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Evaluator.Tests/Org.Apache.REEF.Evaluator.Tests.csproj]: This property can be moved up to the main property group.

File: lang/cs/Org.Apache.REEF.Evaluator.Tests/Properties/AssemblyInfo.cs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+// General Information about an assembly is controlled through the following 
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("Org.Apache.REEF.Evaluator.Tests")]
+[assembly: AssemblyDescription("")]
+[assembly: AssemblyConfiguration("")]
+[assembly: AssemblyCompany("")]
+[assembly: AssemblyProduct("Org.Apache.REEF.Evaluator.Tests")]
+[assembly: AssemblyCopyright("Copyright ©  2015")]
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Evaluator.Tests/Properties/AssemblyInfo.cs]: You can copy the text we use here from another assemblyinfo.cs.

**Comment by markusweimer** [comment]
@jwang98052 I did a pass.

**Comment by markusweimer** [comment]
I just tested it, and the new tests fail:

```
Test Name:	TestEvaluatorConfigurations
Test FullName:	Org.Apache.REEF.Evaluator.Tests.EvaluatorConfigurationsTests.TestEvaluatorConfigurations
Test Source:	c:\src\reef\lang\cs\Org.Apache.REEF.Evaluator.Tests\EvaluatorConfigurationsTests.cs : line 39
Test Outcome:	Failed
Test Duration:	0:00:00.3372919

Result Message:	Assert.IsTrue failed.
Result StackTrace:	at Org.Apache.REEF.Evaluator.Tests.EvaluatorConfigurationsTests.TestEvaluatorConfigurations() in c:\src\reef\lang\cs\Org.Apache.REEF.Evaluator.Tests\EvaluatorConfigurationsTests.cs:line 39
Result StandardOutput:	
Org.Apache.REEF.Common.Runtime.Evaluator.Utils.EvaluatorConfigurations Start: 0 : 2015-07-24T08:32:57.9825254-07:00 0017
START: 2015-07-24 08:32:57 EvaluatorConfigurations::EvaluatorConfigurations
Org.Apache.REEF.Common.Runtime.Evaluator.Utils.EvaluatorConfigurations Stop: 0 : 2015-07-24T08:32:58.2485405-07:00 0017
EXIT: 2015-07-24 08:32:57 EvaluatorConfigurations::EvaluatorConfigurations. Duration: [00:00:00.2663786].
```

File: lang/cs/Org.Apache.REEF.Evaluator.Tests/Org.Apache.REEF.Evaluator.Tests.csproj (hunk 2 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Evaluator.Tests/Org.Apache.REEF.Evaluator.Tests.csproj]: Yeah, if it appears in every one, then we can promote it to build.prop.

File: lang/cs/Org.Apache.REEF.Evaluator.Tests/Org.Apache.REEF.Evaluator.Tests.csproj (hunk 1 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Evaluator.Tests/Org.Apache.REEF.Evaluator.Tests.csproj]: It is relative path added by VS. Should be safe.

File: lang/cs/Org.Apache.REEF.Evaluator.Tests/Org.Apache.REEF.Evaluator.Tests.csproj (hunk 3 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Evaluator.Tests/Org.Apache.REEF.Evaluator.Tests.csproj]: Yes

File: lang/cs/Org.Apache.REEF.Evaluator.Tests/Properties/AssemblyInfo.cs (hunk 4 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Evaluator.Tests/Properties/AssemblyInfo.cs]: I don't see the difference comparing with other assemblyinfo.cs. Could you please specify?

Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Evaluator.Tests/Properties/AssemblyInfo.cs]: Good point. I thought we had changed the copyright and all. We did not :)

**Comment by jwang98052** [comment]
Test issue fixed and review comments addressed and pushed in.

**Comment by markusweimer** [comment]
@jwang98052 Thanks, I will do another test run

**Comment by markusweimer** [comment]
@jwang98052 It passed tetsts before the last commit, but now fails RAT:

```
Unapproved licenses:
  C:/src/reef/lang/cs/Org.Apache.REEF.Examples/ConfigFiles/evaluator.conf
```

**Comment by jwang98052** [comment]
Sorry. Change pushed in.

**Comment by markusweimer** [comment]
@jwang98052 Thanks, I will test and merge.

**Comment by markusweimer** [comment]
After clean, a `Release` build in Visual Studio fails with the following error log:

```
Error	19	Could not find file 'C:\src\reef\lang\cs\Org.Apache.REEF.Evaluator.Tests\Org.Apache.REEF.Evaluator.Tests.nuspec'.	Org.Apache.REEF.Evaluator.Tests
Error	20	The command ""C:\src\reef\lang\cs\.nuget\NuGet.exe" pack "C:\src\reef\lang\cs\\Org.Apache.REEF.Evaluator.Tests\Org.Apache.REEF.Evaluator.Tests.nuspec" -BasePath C:\src\reef\lang\cs\\Org.Apache.REEF.Evaluator.Tests -Properties "Configuration=Release;Platform=x64;REEF_Version=0.12.0-incubating-SNAPSHOT;Version=0.12.0-SNAPSHOT-03" -NonInteractive -OutputDirectory "C:\src\reef\lang\cs\\.nuget\packages" " exited with code 1.	Org.Apache.REEF.Evaluator.Tests
```
A `Debug` build succeeded. Using the `Debug` build, the tests passed. 

Building the `Release` configuration from the command line using

```
msbuild $REEFSourcePath\lang\cs\Org.Apache.REEF.sln /p:Configuration="Release" /p:Platform="x64" /m
```

Also fails:

```
 "c:\src\reef\lang\cs\Org.Apache.REEF.sln" (default target) (1) ->
       "c:\src\reef\lang\cs\Org.Apache.REEF.Evaluator.Tests\Org.Apache.REEF.Evaluator.Tests.csproj" (default target) (16) ->
       (BuildPackage target) ->
         c:\src\reef\lang\cs\.nuget\NuGet.targets(152,9): error : Could not find file 'c:\src\reef\lang\cs\Org.Apache.REEF.Evaluator.Tests\Org.Apache.REEF.Evaluator.Tests.nuspec'. [c:\src\reef\lang\cs\Org.Apache.REEF.Evaluator.Tests\Org.Apache.REEF.Evaluator.Tests.csproj]
         c:\src\reef\lang\cs\.nuget\NuGet.targets(152,9): error MSB3073: The command ""c:\src\reef\lang\cs\.nuget\NuGet.exe" pack "c:\src\reef\lang\cs\\Org.Apache.REEF.Evaluator.Tests\Org.Apache.REEF.Evaluator.Tests.nuspec" -BasePath c:\src\reef\lang\cs\\Org.Apache.REEF.Evaluator.Tests -Properties "Configuration=Relea
       se;Platform=x64;REEF_Version=0.12.0-incubating-SNAPSHOT;Version=0.12.0-SNAPSHOT-03" -NonInteractive -OutputDirectory "c:\src\reef\lang\cs\\.nuget\packages" " exited with code 1. [c:\src\reef\lang\cs\Org.Apache.REEF.Evaluator.Tests\Org.Apache.REEF.Evaluator.Tests.csproj]
```

Hence, I believe this isn't just Visual Studio flakiness but a real issue.

**Comment by markusweimer** [comment]
I believe the issue is that the project indicates to create packages, which makes no sense for a test package. I'll try to fix it while merging.

**Comment by markusweimer** [comment]
I removed the import of the `nuget.targets` at the bottom of the project file. This fixed the issue. I'll test and merge.

**Pull request closed without merging**